src/Dialect/ONNX/ONNXOps/ControlFlow/Loop.cpp

```diff
@@ -42,56 +47,44 @@ LogicalResult ONNXLoopOp::inferShapes(
-  // shape inference to obtain body function output types.
+  // Now we have modified loop body input types according to
+  // the knowledge we have on the initial inputs. Dispatch
+  // shape inference to obtain body output types.
   doShapeInference(loopBody);
+  Operation *terminator = loopBody.back().getTerminator();
+  assert(terminator->getNumOperands() == 1 + getVFinalAndScanOutputs().size());
+  // Skip the termination condition.
+  auto bodyOuputTys = llvm::drop_begin(terminator->getOperandTypes(), 1);
 
-  // Output loop variables should have the same type as their input
-  // counterparts.
-  auto bodyResultTys = loopBody.back().getTerminator()->getOperandTypes();
-  // Compute the type range corresponding to the final values of
-  // loop-carried dependencies/scan outputs in the body function output
-  // types.
-  auto scanStartItr =
-      std::next(bodyResultTys.begin(), 1 + getVInitial().size());
-  auto bodyResVFinalTys =
-      llvm::make_range(std::next(bodyResultTys.begin(), 1), scanStartItr);
-  auto bodyResScanTys = llvm::make_range(scanStartItr, bodyResultTys.end());
-
-  // Set shape for loop operation outputs corresponding to the final
-  // values of loop-carried dependencies to be shape of their counterparts
-  // in the body function output.
-  for (auto vFinalValToTy : llvm::zip(v_final(), bodyResVFinalTys)) {
-    std::get<0>(vFinalValToTy).setType(std::get<1>(vFinalValToTy));
-  }
+  // Set loop carried dependency LoopOp output types (and shapes) to the types
+  // and inferred shapes of their counterparts in the body output.
+  // zip() runs through the shortest range which is v_final().
+  for (auto [val, ty] : llvm::zip(v_final(), bodyOuputTys))
+    val.setType(ty);
 
   // For scan outputs, we set their shape to be the shape of the return
   // values of the loop body function corresponding to scan outputs, but
   // with an extra leading dimension.
-  for (auto vScanOutputValToTy : llvm::zip(scan_outputs(), bodyResScanTys)) {
-    auto rankedScanTy =
-        std::get<1>(vScanOutputValToTy).cast<RankedTensorType>();
-    auto shape = rankedScanTy.getShape();
-    SmallVector<int64_t, 4> unsqueezedShape(shape.begin(), shape.end());
-    // Note that we may know the extent of the scan output leading
-    // dimension, which is very likely just the trip count specified as an
-    // input to Loop operation, but we need to eliminate the possibility of
-    // early termination to be sure.
-    unsqueezedShape.insert(unsqueezedShape.begin(), ShapedType::kDynamic);
-    updateType(std::get<0>(vScanOutputValToTy), unsqueezedShape,
-        rankedScanTy.getElementType(), /*encoding=*/nullptr,
-        /*refineShape=*/false);
+  auto bodyScanOutputTys = llvm::drop_begin(bodyOuputTys, numCarried);
+  for (auto [opScanOutput, ty] : llvm::zip(scan_outputs(), bodyScanOutputTys)) {
+    // TODO: Handle SeqType, OptType.
+    if (auto rankedTy = ty.dyn_cast<RankedTensorType>()) {
```

Comment:
I guess that this IF is the fix to avoid crash on unranked tensor. The PR is more precisely to fix unranked return, isn't it

Reply:
yes, in PR #2098 Loop and Range crashes on unranked return types

the unranked lit tests in this PR also caused a crash in Scan's input scan types